File: player.py
import pygame

from circleshape import CircleShape
from constants import (
    LINE_WIDTH,
    PLAYER_ACCELERATION,
    PLAYER_FRICTION,
    PLAYER_INVULNERABLE_SECONDS,
    PLAYER_MAX_SPEED,
    PLAYER_RADIUS,
    PLAYER_SHOOT_COOLDOWN_SECONDS,
    PLAYER_SHOOT_SPEED,
    PLAYER_TURN_SPEED,
    PLAYER_DASH_SPEED,
    PLAYER_DASH_DURATION_SECONDS,
    PLAYER_DASH_COOLDOWN_SECONDS,
)
from shot import Shot


class Player(CircleShape):
    shoot_sound: pygame.mixer.Sound | None = None
    movement_acceleration: float = PLAYER_ACCELERATION
    movement_max_speed: float = PLAYER_MAX_SPEED
    movement_friction: float = PLAYER_FRICTION

    def __init__(self, x: int, y: int) -> None:
        super().__init__(x, y, PLAYER_RADIUS)
        self.rotation = 0
        self.shoot_timer = 0.0
        self.invulnerable_timer = 0.0
        self.dash_cooldown_timer = 0.0
        self.dash_active_timer = 0.0
        self.dash_direction = pygame.Vector2(0, 0)
        self.dash_trail: list[list[pygame.Vector2]] = []


    def rotate(self, dt: float) -> None:
        self.rotation += PLAYER_TURN_SPEED * dt

    def accelerate(self, dt: float, direction_multiplier: float = 1.0) -> None:
        direction = pygame.Vector2(0, 1).rotate(self.rotation)
        self.velocity += (
            direction * self.movement_acceleration * direction_multiplier * dt
        )

    def move(self, dt: float) -> None:
        self.position += self.velocity * dt

    def apply_friction(self, dt: float) -> None:
        self.velocity *= max(0.0, 1 - self.movement_friction * dt)

    def clamp_speed(self) -> None:
        if self.velocity.length() > self.movement_max_speed:
            self.velocity.scale_to_length(self.movement_max_speed)

    def shoot(self) -> None:
        if self.shoot_timer > 0:
            return

        self.shoot_timer = PLAYER_SHOOT_COOLDOWN_SECONDS
        shot = Shot(self.position.x, self.position.y)
        shot.velocity = pygame.Vector2(0, 1).rotate(self.rotation) * PLAYER_SHOOT_SPEED
        if self.shoot_sound is not None:
            self.shoot_sound.play()

    def _is_key_pressed(
        self,
        keys: pygame.key.ScancodeWrapper | set[int],
        key: int,
    ) -> bool:
        if isinstance(keys, set):
            return key in keys or bool(pygame.key.get_pressed()[key])
        return bool(keys[key])

    def triangle(self) -> list[pygame.Vector2]:
        forward = pygame.Vector2(0, 1).rotate(self.rotation)
        right = pygame.Vector2(0, 1).rotate(self.rotation + 90) * self.radius / 1.5
        a = self.position + forward * self.radius
        b = self.position - forward * self.radius - right
        c = self.position - forward * self.radius + right
        return [a, b, c]

    def dash(self) -> None:
        if self.dash_cooldown_timer > 0:
            return

        self.dash_active_timer = PLAYER_DASH_DURATION_SECONDS
        self.dash_cooldown_timer = PLAYER_DASH_COOLDOWN_SECONDS
        self.dash_direction = pygame.Vector2(0, 1).rotate(self.rotation).normalize()
        self.invulnerable_timer = max(self.invulnerable_timer, PLAYER_DASH_DURATION_SECONDS)
        self.dash_trail.clear()

    def draw(self, screen: pygame.Surface) -> None:
        # Draw trailing dash ghosts
        for points in self.dash_trail:
            pygame.draw.polygon(screen, (80, 80, 80), points, 1)

        if self.is_invulnerable() and pygame.time.get_ticks() % 200 < 100:
            return

        pygame.draw.polygon(
            screen,
            "white",
            self.triangle(),
            LINE_WIDTH,
        )

    def update(
        self, dt: float, keys: pygame.key.ScancodeWrapper | set[int] | None = None
    ) -> None:
        if keys is None:
            keys = pygame.key.get_pressed()

        self.shoot_timer = max(0.0, self.shoot_timer - dt)
        self.invulnerable_timer = max(0.0, self.invulnerable_timer - dt)
        self.dash_cooldown_timer = max(0.0, self.dash_cooldown_timer - dt)

        # Handle active dash
        if self.dash_active_timer > 0:
            self.dash_active_timer = max(0.0, self.dash_active_timer - dt)
            self.velocity = self.dash_direction * PLAYER_DASH_SPEED
            self.move(dt)
            self.wrap_around()
            # Store the current triangle shape for the trail
            self.dash_trail.append(self.triangle())
            if len(self.dash_trail) > 5:
                self.dash_trail.pop(0)
            return

        # Decay trail when not active
        if self.dash_trail:
            self.dash_trail.pop(0)

        if self._is_key_pressed(keys, pygame.K_a):
            self.rotate(-dt)
        if self._is_key_pressed(keys, pygame.K_d):
            self.rotate(dt)

        if self._is_key_pressed(keys, pygame.K_LSHIFT):
            self.dash()

        if self._is_key_pressed(keys, pygame.K_w):
            self.accelerate(dt, 1.0)
        if self._is_key_pressed(keys, pygame.K_s):
            self.accelerate(dt, -1.0)

        self.clamp_speed()
        self.apply_friction(dt)
        self.move(dt)
        self.wrap_around()

        if self._is_key_pressed(keys, pygame.K_SPACE):
            self.shoot()

    def respawn(self, x: float, y: float) -> None:
        self.position = pygame.Vector2(x, y)
        self.rotation = 0
        self.velocity = pygame.Vector2(0, 0)
        self.invulnerable_timer = PLAYER_INVULNERABLE_SECONDS

    def is_invulnerable(self) -> bool:
        return self.invulnerable_timer > 0

File: lib/game/components/player.dart
import 'dart:math' as math;
import 'package:flutter/services.dart';
import 'package:flame/components.dart';
import 'package:flame/collisions.dart';
import 'package:flutter/material.dart';
import 'package:asteroids/constants.dart';
import 'package:asteroids/game/asteroids_game.dart';
import 'package:asteroids/game/weapons.dart';
import 'asteroid.dart';

class PlayerComponent extends PositionComponent with HasGameReference<AsteroidsGame>, CollisionCallbacks {
  Vector2 velocity = Vector2.zero();
  
  double movementAcceleration = GameConstants.playerAcceleration;
  double movementMaxSpeed = GameConstants.playerMaxSpeed;
  double movementFriction = GameConstants.playerFriction;
  
  double radius = GameConstants.playerRadius;
  double shootTimer = 0.0;
  double invulnerableTimer = GameConstants.playerInvulnerableSeconds;
  double dashCooldownTimer = 0.0;
  double dashActiveTimer = 0.0;
  Vector2 dashDirection = Vector2.zero();
  double muzzleFlashTimer = 0.0;
  double muzzleFlashSize = 0.0;

  late Weapon weapon;
  late final Map<String, Weapon> unlockedWeapons;

  PlayerComponent({required Vector2 position}) {
    this.position = position;
    angle = 0.0;
    anchor = Anchor.center;
    size = Vector2.all(radius * 2);
    weapon = Blaster();
    unlockedWeapons = {'blaster': weapon};
  }

  @override
  Future<void> onLoad() async {
    await super.onLoad();
    add(CircleHitbox());
  }

  @override
  void onCollisionStart(Set<Vector2> intersectionPoints, PositionComponent other) {
    super.onCollisionStart(intersectionPoints, other);
    if (other is AsteroidComponent) {
      if (!isInvulnerable()) {
        game.playSfx('player_hit.wav');
        game.triggerScreenShake(duration: 0.4, intensity: 12.0);

        game.spawnExplosionParticles(
          position: position,
          colorChoices: [Colors.cyan, Colors.white, Colors.blue, Colors.grey],
          count: 30,
          minSize: 2.5,
          maxSize: 5.0,
          minSpeed: 80.0,
          maxSpeed: 250.0,
        );

        game.lives -= 1;
        if (game.lives <= 0) {
          game.gameOver();
        } else {
          respawn(game.size.x / 2, game.size.y / 2);
        }
      }
    }
  }

  Vector2 get forward => Vector2(
        math.cos(angle - math.pi / 2),
        math.sin(angle - math.pi / 2),
      );

  void rotate(double dt, double direction) {
    angle += (GameConstants.playerTurnSpeed * math.pi / 180.0) * direction * dt;
  }

  void accelerate(double dt, double directionMultiplier) {
    velocity += forward * movementAcceleration * directionMultiplier * dt;
  }

  void clampSpeed() {
    if (velocity.length > movementMaxSpeed) {
      velocity.normalize();
      velocity.scale(movementMaxSpeed);
    }
  }

  void applyFriction(double dt) {
    velocity.scale(math.exp(-movementFriction * dt));
  }

  void dash() {
    if (dashCooldownTimer > 0) return;

    dashActiveTimer = GameConstants.playerDashDurationSeconds;
    dashCooldownTimer = GameConstants.playerDashCooldownSeconds;
    dashDirection = forward.normalized();
    invulnerableTimer = math.max(invulnerableTimer, GameConstants.playerDashDurationSeconds);
  }

  void respawn(double x, double y) {
    position.setValues(x, y);
    angle = 0.0;
    velocity.setZero();
    invulnerableTimer = GameConstants.playerInvulnerableSeconds;
  }

  bool isInvulnerable() => invulnerableTimer > 0;

  @override
  void update(double dt) {
    super.update(dt);

    shootTimer = math.max(0.0, shootTimer - dt);
    invulnerableTimer = math.max(0.0, invulnerableTimer - dt);
    dashCooldownTimer = math.max(0.0, dashCooldownTimer - dt);
    muzzleFlashTimer = math.max(0.0, muzzleFlashTimer - dt);

    final keys = game.keysPressed;

    // Handle active dash
    if (dashActiveTimer > 0) {
      dashActiveTimer = math.max(0.0, dashActiveTimer - dt);
      velocity = dashDirection * GameConstants.playerDashSpeed;
      position += velocity * dt;
      wrapAround();

      // Spawn dash ghost trail
      game.add(
        DashGhost(
          position: position,
          initialAngle: angle,
          radius: radius,
        ),
      );
      return;
    }

    // Input Handling
    final stick = game.joystick.stickInput;
    if (stick.length > 0.15 && game.joystick.isDragging) {
      // Joystick drag steering & thrust
      final targetAngle = math.atan2(stick.y, stick.x) + math.pi / 2;
      double diff = targetAngle - angle;
      while (diff < -math.pi) diff += 2 * math.pi;
      while (diff > math.pi) diff -= 2 * math.pi;

      final turnAmount = (GameConstants.playerTurnSpeed * math.pi / 180.0) * dt;
      if (diff.abs() < turnAmount) {
        angle = targetAngle;
      } else {
        angle += turnAmount * diff.sign;
      }

      accelerate(dt, stick.length);
    } else {
      // Standard Keyboard Input Handling
      double rotateDirection = 0.0;
      if (keys.contains(LogicalKeyboardKey.arrowLeft) || keys.contains(LogicalKeyboardKey.keyA)) {
        rotateDirection -= 1.0;
      }
      if (keys.contains(LogicalKeyboardKey.arrowRight) || keys.contains(LogicalKeyboardKey.keyD)) {
        rotateDirection += 1.0;
      }
      if (rotateDirection != 0.0) {
        rotate(dt, rotateDirection);
      }

      if (keys.contains(LogicalKeyboardKey.arrowUp) || keys.contains(LogicalKeyboardKey.keyW)) {
        accelerate(dt, 1.0);
      }
      if (keys.contains(LogicalKeyboardKey.arrowDown) || keys.contains(LogicalKeyboardKey.keyS)) {
        accelerate(dt, -1.0);
      }
    }

    if (keys.contains(LogicalKeyboardKey.shiftLeft) || keys.contains(LogicalKeyboardKey.shiftRight)) {
      dash();
    }

    clampSpeed();
    applyFriction(dt);
    position += velocity * dt;
    wrapAround();

    if (keys.contains(LogicalKeyboardKey.space)) {
      shoot();
    }

    // Weapon switching controls
    if (keys.contains(LogicalKeyboardKey.digit1) && unlockedWeapons.containsKey('blaster')) {
      setWeapon(unlockedWeapons['blaster']!);
    }
    if (keys.contains(LogicalKeyboardKey.digit2) && unlockedWeapons.containsKey('spread_shot')) {
      setWeapon(unlockedWeapons['spread_shot']!);
    }
    if (keys.contains(LogicalKeyboardKey.digit3) && unlockedWeapons.containsKey('rapid_fire')) {
      setWeapon(unlockedWeapons['rapid_fire']!);
    }
    if (keys.contains(LogicalKeyboardKey.digit4) && unlockedWeapons.containsKey('bomb_launcher')) {
      setWeapon(unlockedWeapons['bomb_launcher']!);
    }
  }

  void shoot() {
    if (shootTimer > 0) return;

    if (weapon.fire(this)) {
      shootTimer = weapon.cooldown;
      // Apply physical recoil pushing the player backwards
      velocity -= forward * weapon.recoil;

      // Set muzzle flash visual effect
      muzzleFlashTimer = 0.08;
      muzzleFlashSize = weapon.muzzleFlashSize;

      game.playSfx('shoot.wav');
    }
  }

  void setWeapon(Weapon newWeapon) {
    weapon = newWeapon;
    final key = newWeapon.name.toLowerCase().replaceAll(' ', '_');
    unlockedWeapons[key] = newWeapon;
  }

  void wrapAround() {
    final sizeX = game.size.x;
    final sizeY = game.size.y;

    if (position.x < -radius) {
      position.x = sizeX + radius;
    } else if (position.x > sizeX + radius) {
      position.x = -radius;
    }

    if (position.y < -radius) {
      position.y = sizeY + radius;
    } else if (position.y > sizeY + radius) {
      position.y = -radius;
    }
  }

  @override
  void render(Canvas canvas) {
    // Blinking effect when invulnerable
    if (isInvulnerable()) {
      final milliseconds = DateTime.now().millisecondsSinceEpoch;
      if (milliseconds % 200 < 100) {
        return; // Skip rendering
      }
    }

    canvas.translate(radius, radius);

    final paint = Paint()
      ..color = Colors.white
      ..style = PaintingStyle.stroke
      ..strokeWidth = GameConstants.lineWidth;

    final path = Path()
      ..moveTo(0, -radius)
      ..lineTo(-radius / 1.5, radius)
      ..lineTo(radius / 1.5, radius)
      ..close();

    canvas.drawPath(path, paint);

    // Muzzle flash rendering
    if (muzzleFlashTimer > 0) {
      final flashPaint = Paint()..color = Colors.yellow;
      canvas.drawCircle(Offset(0, -radius), muzzleFlashSize, flashPaint);
      
      final innerFlashPaint = Paint()..color = Colors.white;
      canvas.drawCircle(Offset(0, -radius), muzzleFlashSize * 0.5, innerFlashPaint);
    }
  }
}

class DashGhost extends PositionComponent {
  final double maxLifespan = 0.15;
  double lifespan = 0.15;
  final double radius;

  DashGhost({
    required Vector2 position,
    required double initialAngle,
    required this.radius,
  }) {
    this.position = position.clone();
    angle = initialAngle;
    anchor = Anchor.center;
    size = Vector2.all(radius * 2);
  }

  @override
  void update(double dt) {
    super.update(dt);
    lifespan -= dt;
    if (lifespan <= 0) {
      removeFromParent();
    }
  }

  @override
  void render(Canvas canvas) {
    final alpha = (lifespan / maxLifespan * 80).clamp(0.0, 255.0).toInt();
    if (alpha <= 0) return;

    canvas.translate(radius, radius);

    final paint = Paint()
      ..color = Color.fromARGB(alpha, 255, 255, 255)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1.0;

    final path = Path()
      ..moveTo(0, -radius)
      ..lineTo(-radius / 1.5, radius)
      ..lineTo(radius / 1.5, radius)
      ..close();

    canvas.drawPath(path, paint);
  }
}
